extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SimpleScheduler.java

```diff
@@ -288,29 +292,42 @@ public Trigger getScheduledJob(String identity) {
             this.invoker = invoker;
         }
 
-        void execute(ZonedDateTime now, ExecutorService executor) {
+        void execute(ZonedDateTime now, ExecutorService executor, Vertx vertx) {
             if (!trigger.isRunning()) {
                 return;
             }
             ZonedDateTime scheduledFireTime = trigger.evaluate(now);
             if (scheduledFireTime != null) {
-                try {
-                    executor.execute(new Runnable() {
-                        @Override
-                        public void run() {
-                            try {
-                                invoker.invoke(new SimpleScheduledExecution(now, scheduledFireTime, trigger));
-                            } catch (Throwable t) {
-                                // already logged by the StatusEmitterInvoker
+                if (invoker.isBlocking()) {
+                    try {
+                        executor.execute(new Runnable() {
+                            @Override
+                            public void run() {
+                                doInvoke(now, scheduledFireTime);
                             }
+                        });
+                    } catch (RejectedExecutionException e) {
+                        LOG.warnf("Rejected execution of a scheduled task for trigger %s", trigger);
+                    }
+                } else {
+                    vertx.runOnContext(new Handler<Void>() {
```

Comment:
That's where we need to introduce the duplicate context part. As you already have the vert.x instance, it is straightforward.

Reply:
I see. But I have to admit that I'm still very puzzled with all this duplicated context thingy...